In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv("weatherHistory.csv")
df.head(20)

Formatted Date        Summary Precip Type  Temperature (C)  \
0   2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1   2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2   2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3   2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4   2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   
5   2006-04-01 05:00:00.000 +0200  Partly Cloudy        rain         9.222222   
6   2006-04-01 06:00:00.000 +0200  Partly Cloudy        rain         7.733333   
7   2006-04-01 07:00:00.000 +0200  Partly Cloudy        rain         8.772222   
8   2006-04-01 08:00:00.000 +0200  Partly Cloudy        rain        10.822222   
9   2006-04-01 09:00:00.000 +0200  Partly Cloudy        rain        13.772222   
10  2006-04-01 10:00:00.000 +0200  Partly Cloudy        rain        16.016667   
11  2006-04-01 11:00:00.000 +0200  Partly Cloudy        rain        17.144444   
12  2006-04-01 12:00:00.000 +0200  Partly Cloudy        rain        17.800000   
13  2006-04-01 13:00:00.000 +0200  Partly Cloudy        rain        17.333333   
14  2006-04-01 14:00:00.000 +0200  Partly Cloudy        rain        18.877778   
15  2006-04-01 15:00:00.000 +0200  Partly Cloudy        rain        18.911111   
16  2006-04-01 16:00:00.000 +0200  Partly Cloudy        rain        15.388889   
17  2006-04-01 17:00:00.000 +0200  Mostly Cloudy        rain        15.550000   
18  2006-04-01 18:00:00.000 +0200  Mostly Cloudy        rain        14.255556   
19  2006-04-01 19:00:00.000 +0200  Mostly Cloudy        rain        13.144444   

    Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                   7.388889      0.89            14.1197   
1                   7.227778      0.86            14.2646   
2                   9.377778      0.89             3.9284   
3                   5.944444      0.83            14.1036   
4                   6.977778      0.83            11.0446   
5                   7.111111      0.85            13.9587   
6                   5.522222      0.95            12.3648   
7                   6.527778      0.89            14.1519   
8                  10.822222      0.82            11.3183   
9                  13.772222      0.72            12.5258   
10                 16.016667      0.67            17.5651   
11                 17.144444      0.54            19.7869   
12                 17.800000      0.55            21.9443   
13                 17.333333      0.51            20.6885   
14                 18.877778      0.47            15.3755   
15                 18.911111      0.46            10.4006   
16                 15.388889      0.60            14.4095   
17                 15.550000      0.63            11.1573   
18                 14.255556      0.69             8.5169   
19                 13.144444      0.70             7.6314   

    Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                    251.0          15.8263         0.0               1015.13   
1                    259.0          15.8263         0.0               1015.63   
2                    204.0          14.9569         0.0               1015.94   
3                    269.0          15.8263         0.0               1016.41   
4                    259.0          15.8263         0.0               1016.51   
5                    258.0          14.9569         0.0               1016.66   
6                    259.0           9.9820         0.0               1016.72   
7                    260.0           9.9820         0.0               1016.84   
8                    259.0           9.9820         0.0               1017.37   
9                    279.0           9.9820         0.0               1017.22   
10                   290.0          11.2056         0.0               1017.42   
11                   316.0          11.4471         0.0        

# Analysis Data

In [4]:
df.shape

(96453, 12)

In [6]:
df.index

RangeIndex(start=0, stop=96453, step=1)

In [8]:
df.dtypes

Formatted Date               object
Summary                      object
Precip Type                  object
Temperature (C)             float64
Apparent Temperature (C)    float64
Humidity                    float64
Wind Speed (km/h)           float64
Wind Bearing (degrees)      float64
Visibility (km)             float64
Loud Cover                  float64
Pressure (millibars)        float64
Daily Summary                object
dtype: object

In [12]:
df['Precip Type'].unique()

array(['rain', 'snow', nan], dtype=object)

In [13]:
df.count()

Formatted Date              96453
Summary                     96453
Precip Type                 95936
Temperature (C)             96453
Apparent Temperature (C)    96453
Humidity                    96453
Wind Speed (km/h)           96453
Wind Bearing (degrees)      96453
Visibility (km)             96453
Loud Cover                  96453
Pressure (millibars)        96453
Daily Summary               96453
dtype: int64

In [16]:
df["Precip Type"].value_counts()

rain    85224
snow    10712
Name: Precip Type, dtype: int64

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Formatted Date            96453 non-null  object 
 1   Summary                   96453 non-null  object 
 2   Precip Type               95936 non-null  object 
 3   Temperature (C)           96453 non-null  float64
 4   Apparent Temperature (C)  96453 non-null  float64
 5   Humidity                  96453 non-null  float64
 6   Wind Speed (km/h)         96453 non-null  float64
 7   Wind Bearing (degrees)    96453 non-null  float64
 8   Visibility (km)           96453 non-null  float64
 9   Loud Cover                96453 non-null  float64
 10  Pressure (millibars)      96453 non-null  float64
 11  Daily Summary             96453 non-null  object 
dtypes: float64(8), object(4)
memory usage: 8.8+ MB


# Unique wind speed values in the data


In [19]:
df.head(20)

Formatted Date        Summary Precip Type  Temperature (C)  \
0   2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1   2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2   2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3   2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4   2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   
5   2006-04-01 05:00:00.000 +0200  Partly Cloudy        rain         9.222222   
6   2006-04-01 06:00:00.000 +0200  Partly Cloudy        rain         7.733333   
7   2006-04-01 07:00:00.000 +0200  Partly Cloudy        rain         8.772222   
8   2006-04-01 08:00:00.000 +0200  Partly Cloudy        rain        10.822222   
9   2006-04-01 09:00:00.000 +0200  Partly Cloudy        rain        13.772222   
10  2006-04-01 10:00:00.000 +0200  Partly Cloudy        rain        16.016667   
11  2006-04-01 11:00:00.000 +0200  Partly Cloudy        rain        17.144444   
12  2006-04-01 12:00:00.000 +0200  Partly Cloudy        rain        17.800000   
13  2006-04-01 13:00:00.000 +0200  Partly Cloudy        rain        17.333333   
14  2006-04-01 14:00:00.000 +0200  Partly Cloudy        rain        18.877778   
15  2006-04-01 15:00:00.000 +0200  Partly Cloudy        rain        18.911111   
16  2006-04-01 16:00:00.000 +0200  Partly Cloudy        rain        15.388889   
17  2006-04-01 17:00:00.000 +0200  Mostly Cloudy        rain        15.550000   
18  2006-04-01 18:00:00.000 +0200  Mostly Cloudy        rain        14.255556   
19  2006-04-01 19:00:00.000 +0200  Mostly Cloudy        rain        13.144444   

    Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                   7.388889      0.89            14.1197   
1                   7.227778      0.86            14.2646   
2                   9.377778      0.89             3.9284   
3                   5.944444      0.83            14.1036   
4                   6.977778      0.83            11.0446   
5                   7.111111      0.85            13.9587   
6                   5.522222      0.95            12.3648   
7                   6.527778      0.89            14.1519   
8                  10.822222      0.82            11.3183   
9                  13.772222      0.72            12.5258   
10                 16.016667      0.67            17.5651   
11                 17.144444      0.54            19.7869   
12                 17.800000      0.55            21.9443   
13                 17.333333      0.51            20.6885   
14                 18.877778      0.47            15.3755   
15                 18.911111      0.46            10.4006   
16                 15.388889      0.60            14.4095   
17                 15.550000      0.63            11.1573   
18                 14.255556      0.69             8.5169   
19                 13.144444      0.70             7.6314   

    Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                    251.0          15.8263         0.0               1015.13   
1                    259.0          15.8263         0.0               1015.63   
2                    204.0          14.9569         0.0               1015.94   
3                    269.0          15.8263         0.0               1016.41   
4                    259.0          15.8263         0.0               1016.51   
5                    258.0          14.9569         0.0               1016.66   
6                    259.0           9.9820         0.0               1016.72   
7                    260.0           9.9820         0.0               1016.84   
8                    259.0           9.9820         0.0               1017.37   
9                    279.0           9.9820         0.0               1017.22   
10                   290.0          11.2056         0.0               1017.42   
11                   316.0          11.4471         0.0        

In [22]:
df["Wind Speed (km/h)"].unique()

array([14.1197, 14.2646,  3.9284, ..., 37.0622, 35.5971, 30.751 ])

In [ ]:
#Numer of time when the wether is clear

In [23]:
df["Daily Summary"].value_counts()

Mostly cloudy throughout the day.                                                            20085
Partly cloudy throughout the day.                                                             9981
Partly cloudy until night.                                                                    6169
Partly cloudy starting in the morning.                                                        5184
Foggy in the morning.                                                                         4201
                                                                                             ...  
Mostly cloudy until evening and breezy in the afternoon.                                        24
Mostly cloudy until night and breezy starting in the evening continuing until night.            24
Partly cloudy starting overnight continuing until night and breezy in the afternoon.            24
Mostly cloudy until evening and windy starting in the morning continuing until afternoon.       24
Foggy star

In [38]:
#Filtering Data
df[df.Summary=='Clear']


Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [45]:
df[df["Wind Speed (km/h)"] == 14.00 ]

Empty DataFrame
Columns: [Formatted Date, Summary, Precip Type, Temperature (C), Apparent Temperature (C), Humidity, Wind Speed (km/h), Wind Bearing (degrees), Visibility (km), Loud Cover, Pressure (millibars), Daily Summary]
Index: []

In [107]:
#Null values

df.isnull().sum()

Formatted Date                0
Summary                       0
Precip Type                 517
Temperature (C)               0
Apparent Temperature (C)      0
Humidity                      0
Wind_Speed_km/h               0
Wind Bearing (degrees)        0
Visibility_km                 0
Loud Cover                    0
Pressure (millibars)          0
Daily Wether                  0
dtype: int64

In [49]:
df.notnull().sum()

Formatted Date              96453
Summary                     96453
Precip Type                 95936
Temperature (C)             96453
Apparent Temperature (C)    96453
Humidity                    96453
Wind Speed (km/h)           96453
Wind Bearing (degrees)      96453
Visibility (km)             96453
Loud Cover                  96453
Pressure (millibars)        96453
Daily Summary               96453
dtype: int64

In [76]:
#Rename the column name 'Daily Summary' to "Daily Wether"
df.rename(columns={"Daily Summary":'Daily Wether'},inplace=True)
df.rename(columns={"Wind Speed (km/h)":"Wind_Speed_km/h"},inplace=True)
df.rename(columns={"Visibility (km)":"Visibility_km"},inplace=True)

In [77]:
df.head(2)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   

   Apparent Temperature (C)  Humidity  Wind_Speed_km/h  \
0                  7.388889      0.89          14.1197   
1                  7.227778      0.86          14.2646   

   Wind Bearing (degrees)  Visibility_km  Loud Cover  Pressure (millibars)  \
0                   251.0        15.8263         0.0               1015.13   
1                   259.0        15.8263         0.0               1015.63   

                        Daily Wether  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.

In [82]:
#mean of Visibility and Wind Speed
df.Visibility_km.mean()
print("Visibility_km Mean:-",df.Visibility_km.mean())


Visibility_km Mean:- 10.347324929237148


In [84]:
df.describe()

Temperature (C)  Apparent Temperature (C)      Humidity  \
count     96453.000000              96453.000000  96453.000000   
mean         11.932678                 10.855029      0.734899   
std           9.551546                 10.696847      0.195473   
min         -21.822222                -27.716667      0.000000   
25%           4.688889                  2.311111      0.600000   
50%          12.000000                 12.000000      0.780000   
75%          18.838889                 18.838889      0.890000   
max          39.905556                 39.344444      1.000000   

       Wind_Speed_km/h  Wind Bearing (degrees)  Visibility_km  Loud Cover  \
count     96453.000000            96453.000000   96453.000000     96453.0   
mean         10.810640              187.509232      10.347325         0.0   
std           6.913571              107.383428       4.192123         0.0   
min           0.000000                0.000000       0.000000         0.0   
25%           5.828200              116.000000       8.339800         0.0   
50%           9.965900              180.000000      10.046400         0.0   
75%          14.135800              290.000000      14.812000         0.0   
max          63.852600              359.000000      16.100000         0.0   

       Pressure (millibars)  
count          96453.000000  
mean            1003.235956  
std              116.969906  
min                0.000000  
25%             1011.900000  
50%             1016.450000  
75%             1021.090000  
max             1046.380000

In [ ]:
#IF space between two word like"Loud Cover"then we use Square Breckets[],neighter " . "

In [85]:
df["Humidity"].var()

0.03820959171844407

In [87]:
df["Wind_Speed_km/h"].var()

47.79746414615446

In [88]:
df.head(2)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   

   Apparent Temperature (C)  Humidity  Wind_Speed_km/h  \
0                  7.388889      0.89          14.1197   
1                  7.227778      0.86          14.2646   

   Wind Bearing (degrees)  Visibility_km  Loud Cover  Pressure (millibars)  \
0                   251.0        15.8263         0.0               1015.13   
1                   259.0        15.8263         0.0               1015.63   

                        Daily Wether  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.

In [97]:
df[df["Precip Type"]== 'Snow']



Empty DataFrame
Columns: [Formatted Date, Summary, Precip Type, Temperature (C), Apparent Temperature (C), Humidity, Wind_Speed_km/h, Wind Bearing (degrees), Visibility_km, Loud Cover, Pressure (millibars), Daily Wether]
Index: []

In [106]:
df.Summary.str.contains('Snow').head

<bound method NDFrame.head of 0        False
1        False
2        False
3        False
4        False
         ...  
96448    False
96449    False
96450    False
96451    False
96452    False
Name: Summary, Length: 96453, dtype: bool>

In [114]:
df.Summary

0        Partly Cloudy
1        Partly Cloudy
2        Mostly Cloudy
3        Partly Cloudy
4        Mostly Cloudy
             ...      
96448    Partly Cloudy
96449    Partly Cloudy
96450    Partly Cloudy
96451    Partly Cloudy
96452    Partly Cloudy
Name: Summary, Length: 96453, dtype: object

In [115]:
import numpy as np


In [117]:
df.replace("?", np.nan, inplace = True)# replace "?" to NaN


In [120]:
missing_data=df.isnull()
missing_data.head()


Formatted Date  Summary  Precip Type  Temperature (C)  \
0           False    False        False            False   
1           False    False        False            False   
2           False    False        False            False   
3           False    False        False            False   
4           False    False        False            False   

   Apparent Temperature (C)  Humidity  Wind_Speed_km/h  \
0                     False     False            False   
1                     False     False            False   
2                     False     False            False   
3                     False     False            False   
4                     False     False            False   

   Wind Bearing (degrees)  Visibility_km  Loud Cover  Pressure (millibars)  \
0                   False          False       False                 False   
1                   False          False       False                 False   
2                   False          False       False                 False   
3                   False          False       False                 False   
4                   False          False       False                 False   

   Daily Wether  
0         False  
1         False  
2         False  
3         False  
4         False

#"True" stands for missing value, while "False" stands for not missing value.

#Count missing values in each column
#Using a for loop in Python, we can quickly figure out the number of missing values in each column. 
#As mentioned above, "True" represents a missing value, "False" means the value is present in the dataset.
#In the body of the for loop the method ".value_counts()" counts the number of "True" values.

In [121]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")   

Formatted Date
False    96453
Name: Formatted Date, dtype: int64

Summary
False    96453
Name: Summary, dtype: int64

Precip Type
False    95936
True       517
Name: Precip Type, dtype: int64

Temperature (C)
False    96453
Name: Temperature (C), dtype: int64

Apparent Temperature (C)
False    96453
Name: Apparent Temperature (C), dtype: int64

Humidity
False    96453
Name: Humidity, dtype: int64

Wind_Speed_km/h
False    96453
Name: Wind_Speed_km/h, dtype: int64

Wind Bearing (degrees)
False    96453
Name: Wind Bearing (degrees), dtype: int64

Visibility_km
False    96453
Name: Visibility_km, dtype: int64

Loud Cover
False    96453
Name: Loud Cover, dtype: int64

Pressure (millibars)
False    96453
Name: Pressure (millibars), dtype: int64

Daily Wether
False    96453
Name: Daily Wether, dtype: int64



Based on the summary above, each column has 96453 rows of data, Precip Type
 columns containing 517 missing data:

Deal with missing data
#drop data
a. drop the whole row
b. drop the whole column
#replace data
a. replace it by mean
b. replace it by frequency
c. replace it based on other functions

In [139]:
df.groupby('Precip Type').max()


Formatted Date                  Summary  \
Precip Type                                                           
rain         2016-12-31 20:00:00.000 +0100  Windy and Partly Cloudy   
snow         2016-12-31 23:00:00.000 +0100       Windy and Overcast   

             Temperature (C)  Apparent Temperature (C)  Humidity  \
Precip Type                                                        
rain               39.905556                 39.344444       1.0   
snow                0.000000                  0.000000       1.0   

             Wind_Speed_km/h  Wind Bearing (degrees)  Visibility_km  \
Precip Type                                                           
rain                 63.8526                   359.0           16.1   
snow                 44.8868                   359.0           16.1   

             Loud Cover  Pressure (millibars)                Daily Wether  
Precip Type                                                                
rain                0.0               1042.38     Windy in the afternoon.  
snow                0.0               1046.38  Partly cloudy until night.

In [140]:
df["Precip Type"].value_counts()

rain    85224
snow    10712
Name: Precip Type, dtype: int64

#In this "Precip Type" column counting of 'Rain' is more than 'Snow' ,i.e 
#replce Nan to Rain

In [143]:
df["Precip Type"].replace(np.nan, 'rain' , inplace = True)


In [144]:
df.isnull().sum()

Formatted Date              0
Summary                     0
Precip Type                 0
Temperature (C)             0
Apparent Temperature (C)    0
Humidity                    0
Wind_Speed_km/h             0
Wind Bearing (degrees)      0
Visibility_km               0
Loud Cover                  0
Pressure (millibars)        0
Daily Wether                0
dtype: int64

Correct data format
We are almost there!

In [145]:
df.dtypes

Formatted Date               object
Summary                      object
Precip Type                  object
Temperature (C)             float64
Apparent Temperature (C)    float64
Humidity                    float64
Wind_Speed_km/h             float64
Wind Bearing (degrees)      float64
Visibility_km               float64
Loud Cover                  float64
Pressure (millibars)        float64
Daily Wether                 object
dtype: object

In [146]:
df.shape

(96453, 12)

#Wonderful!

Now, we finally obtain the cleaned dataset with no missing values and all data in its proper format.

In [148]:
df.head(5)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind_Speed_km/h  \
0                  7.388889      0.89          14.1197   
1                  7.227778      0.86          14.2646   
2                  9.377778      0.89           3.9284   
3                  5.944444      0.83          14.1036   
4                  6.977778      0.83          11.0446   

   Wind Bearing (degrees)  Visibility_km  Loud Cover  Pressure (millibars)  \
0                   251.0        15.8263         0.0               1015.13   
1                   259.0        15.8263         0.0               1015.63   
2                   204.0        14.9569         0.0               1015.94   
3                   269.0        15.8263         0.0               1016.41   
4                   259.0        15.8263         0.0               1016.51   

                        Daily Wether  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [6]:
%matplotlib inline
import matplotlib as plt
from matplotlib import pyplot
import pandas as pd


In [ ]:
df=